client/src/extension.ts

```diff
@@ -116,16 +118,85 @@ export function activate(context: ExtensionContext) {
         // nothing more to do
         return;
     }
+
+    // First check recursively all the files set in `project.kao`-like files, starting with `project.kao`.
+    workspace.findFiles(`**/project.kao`).then((uris) => {
+        uris.forEach((uri) => {
+            if (openProjectFile(uri)) {
+                // There should be only one `project.kao` file. If we found a good candidate, stop the loop.
+                return;
+            }
+        });
+    });
+
     /*
-     * List of all the yseopml file extensions known by this extension as set in `client/package.json`.
+     * List of all the yseopml file extensions known by this extension as set in `client/package.json`
+     * but never explictly set in project.kao-like files.
      */
-    const yseopmlExtensions = ['kao', 'yclass', 'yobject', 'ycomplete', 'dcl', 'yml'];
-
+    const yseopmlExtensions = ['yclass', 'yobject', 'ycomplete'];
     for (const extension of yseopmlExtensions) {
         parseFilesWithExtension(extension);
     }
 }
 
+/**
+ * Try to open a file with URI `fileUri`.
+ * Then, if the file is a `project.kao`-like file (i.e. is a list files used for the project),
+ * read its content and apply this function recursivly for each line that are existing file's URI.
+ *
+ * @param fileUri An existing file URI.
+ *
+ * @return `true` only if the provided URI was a `*.kao`-like file i.e. its first line starts with `_FILE_TYPE_`.
+ */
+function openProjectFile(fileUri: Uri): boolean {
+    let wasKaoFile = true;
+
+    // Try to open the file. If it is opened, the server will parse it.
+    workspace.openTextDocument(fileUri).then(
+        // The document exists and was successfully opened and should be parsed already.
+        (doc) => {
+            console.debug(`Parsing ${fileUri}.`);
+            const lines = doc.getText().split('\n');
+            if (lines.length === 0 || !lines[0].trim().startsWith('_FILE_TYPE_')) {
+                // We are not in a `project.kao`-like file. Do not go further.
+                wasKaoFile = false;
+                return;
+            }
+            lines
+                // line can be indented in the file.
+                .map((line) => line.trim())
+                /*
+                 * Ignore:
+                 * - any empty line
```

Comment:
* Again, I’d move each of these comments next to the corresponding test.
* I’d use a succession of filters rather than a single obese one.
* Aren’t you recompiling the same regex over and over again, here?
* As far as I know, `/* …` is more common than `//` in those files. Wouldn’t hurt to skip that too. Sure, we won’t be able to tell how many lines the comment spans, but at least we will be skipping the first one of each (which is often the only one, as in the typical `/** end of project specific beforeExternReferences.dcl **/` lines).

```ts
// Empty lines.
.filter((line) => line.length > 0)
// Preprocessing instructions.
.filter((line) => !line.startsWith('@'))
// File type header.
.filter((line) => !line.startsWith('_FILE_TYPE_'))
// Comments.
.filter((line) => !line.startsWith('//'))
.filter((line) => !line.startsWith('/*'))
// Files in “generated YML” directories.
.filter((line) => line.search(BLAHBLAH_CONSTANT_PRE_COMPILED_REGEX)/) === -1)
```

(Did what I could with my limited knowledge xD)

Reply:
> * Again, I’d move each of these comments next to the corresponding test.

OK!

> * I’d use a succession of filters rather than a single obese one.

In fact, each call to `filter` does the following: `Returns the elements of an array that meet the condition specified in a callback function.`. I don't want to have a new array created for each small piece of condition. Don't forget that we don't have here a compilator that will increase code performance. There is no JAVA-like `Stream` neither.

> * Aren’t you recompiling the same regex over and over again, here?

There is a high probability here, yeah. Good catch.

> * As far as I know, /* … is more common than // in those files. Wouldn’t hurt to skip that too. Sure, we won’t be able to tell how many lines the comment spans, but at least we will be skipping the first one of each (which is often the only one, as in the typical /** end of project specific beforeExternReferences.dcl **/ lines).

Well, I never see `the typical […] lines` in any of my projects.
But if you say it's that important, I'll add these tests.

Reply:
OK if you think the multiple filters will decrease perfs.

> Well, I never see the typical […] lines in any of my projects.
> But if you say it's that important, I'll add these tests.

```bash
$ locate -0 project.kao | xargs -0 grep -ho '^\s*/[/*]' | sed 's/^\s*//' | sort | uniq -c
      4 //
    410 /*
```

I think those numbers of occurrences speak for themselves.